_posts/2020-12-21-multipleSchemas.md

```diff
@@ -0,0 +1,118 @@
+```
+
+```
+#!/bin/bash
+schemas=(foo bar)
+for schema in "${schemas[@]}"
+do
+    flyway migrate -configFiles="/src/config/$schema.conf" 
+done
+```
+
+## One desired schema structure, many histories
+
+Now suppose we have a multi-tenant database where we want all schemas to have an identical structure. Now, we no longer
+need separate sets of scripts - and moreover, we no longer want separate sets of scripts, as having a single set will
+enforce that uniformity for us. Instead of iterating over configurations, we can iterate instead over the schemas,
+using the `${flyway:defaultSchema}` placeholder in any script that needs a fully qualified name:
+
+```
+# Powershell
+$schemas = @('foo', 'bar')
+foreach ($schema in $schemas) { flyway migrate -configfFiles="c:\src\config\flyway.conf" -defaultSchema="$schema" }
+```
+
+Having a separate history table in each schema appears redundant in that the same sequence of migrations will be
+recorded in each table, and calling `migrate` repeatedly like this will migrate each schema up to the same state
+(assuming no new migrations are created while Flyway is busy!). However, it does mean that when we need a new
+schema with the same structure, we can simply add it
+to the list, and Flyway will create it automatically and migrate it forwards through the same sequence as all the
+others until it has reached the latest structure.
+
+This method is also failure-resistant - if Flyway fails on one schema, the history tables contain enough information
+that once the failure is cleared, Flyway will do the right thing on future runs.  
+
+It is possible to do similar iterations with Maven and Gradle; in addition in Maven there is the 
+[`iterator-maven-plugin`](https://github.com/khmarbaise/iterator-maven-plugin) which makes iterating over a collection of 
+
+## One desired schema structure, one history
+
+This is the most tricky scenario, as it requires each migration to handle the iteration across schemas - which makes
+the migrations more complex as they will need to use dynamic SQL. Adding a new schema is also much more complex;
+suppose we've run three migrations on schemas `foo` and `bar`, and now we want to add schema `baz`. The history table
+tells us we've run three migrations, but not which schemas were involved. So to get `baz` up to the correct state
+we need another migration, #4 - and that migration now needs to check each schema in turn, and if they are not at the
+right state, re-run migrations #1, #2 and #3 somehow. 
+
+The prospect of errors during the catch-up process complicates things further. What happens if something fails
+while schema `baz` is being brought up to the current state? We have no way of reflecting that in the history table as 
+it stands, and so we would need more SQL to figure out what state it was in. 
+
+So, we can't recommend this approach.
+
+## But that's not the end of the story...
```

Comment:
I wonder if we can preface this blog with a similar call to action. Something to the effect of 'We've had some users ask about multi-schema deployments. Flyway supports some scenarios but not others. This blog post is our current thinking on the matter.  If you have any feedback for us, post in this GitHub issue.'